processing/src/main/java/org/apache/druid/query/filter/ColumnIndexSelector.java

```diff
@@ -19,33 +19,33 @@
 
 package org.apache.druid.query.filter;
 
-import com.google.errorprone.annotations.MustBeClosed;
 import org.apache.druid.collections.bitmap.BitmapFactory;
-import org.apache.druid.collections.bitmap.ImmutableBitmap;
-import org.apache.druid.collections.spatial.ImmutableRTree;
 import org.apache.druid.segment.ColumnInspector;
-import org.apache.druid.segment.column.BitmapIndex;
-import org.apache.druid.segment.column.ColumnCapabilities;
-import org.apache.druid.segment.data.CloseableIndexed;
+import org.apache.druid.segment.column.ColumnIndexCapabilities;
 
 import javax.annotation.Nullable;
 
 /**
  */
-public interface BitmapIndexSelector extends ColumnInspector
+public interface ColumnIndexSelector extends ColumnInspector
 {
-  @MustBeClosed
-  @Nullable
-  CloseableIndexed<String> getDimensionValues(String dimension);
-
-  @Deprecated
-  ColumnCapabilities.Capable hasMultipleValues(String dimension);
-
   int getNumRows();
+
   BitmapFactory getBitmapFactory();
+
+  /**
+   * Get the {@link ColumnIndexCapabilities} of a column for the specified type of index. If the index does not exist
+   * this method will return null. Note that 'missing' columns should in fact return a non-null value from this method
+   * to allow for filters to use 'nil' bitmaps if the filter matches nulls, in order to produce an all true or all
+   * false index.
+   */
   @Nullable
-  BitmapIndex getBitmapIndex(String dimension);
+  <T> ColumnIndexCapabilities getIndexCapabilities(String column, Class<T> clazz);
+
+  /**
+   * Get the specified type of index for the specified column. {@link #getIndexCapabilities(String, Class)} should
+   * be called prior to this method to distinguish 'missing' columns from columns without indexes.
+   */
   @Nullable
-  ImmutableBitmap getBitmapIndex(String dimension, String value);
-  ImmutableRTree getSpatialIndex(String dimension);
+  <T> T as(String column, Class<T> clazz);
```

Comment:
This makes sense to just expose the index supplier here, and eliminates the need to check capabilities as a proxy for if the column exists or not. However, I think we still need index capabilities for other reasons to provide information about characteristics of the indexes themselves. It would need to return a non-null index supplier for all columns that exist, and that index supplier can just return nulls for capabilities and indexes, which is easy enough.

Reply:
For the capabilities, what I was suggesting is, e.g. where LexicographicRangeIndex has method

```
default Iterable<ImmutableBitmap> getBitmapsInRange(
      @Nullable String startValue,
      boolean startStrict,
      @Nullable String endValue,
      boolean endStrict
  )
```

It could have signature

```
default FilterBundle getBitmapsInRange(
      @Nullable String startValue,
      boolean startStrict,
      @Nullable String endValue,
      boolean endStrict
  )
```

Where `FilterBundle` would be a new class that could have fields for bitmap indexes, value predicates, selectivity estimation, whether it's safe to invert, etc. etc.  Which, I believe meets the need that you are highlighting for `ColumnIndexCapabilities`.

Reply:
I think `FilterBundle` is going to be a bit more complex than you're hoping, which is part of why I didn't try to consolidate `getIndexCapabilities` and `getIndex` yet. Not all value matchers are predicates, so it can't strictly operate on predicates, and value matchers need the cursor created from the offset of the 'pre' part of the filter bundle to actually be created, so it might be a factory that can make value matchers to use as a post-filter.